libs/state/src/lib/transformation-helpers/array/upsert.ts

```diff
@@ -0,0 +1,48 @@
+import { isDefined } from '../../core/utils/guards';
+import { valuesComparer } from '../_internals/valuesComparer.util';
+import { ComparableData } from '../interfaces/comparable-data-type';
+
+export function upsert<T extends object>(
```

Comment:
I already thought about this but came to the conclusion that this would ultimately make things less performant. Upsert would have to do all the edge case checks (like valid input, is source of correct type, ...etc.) before it can decide to insert and upsert - where all of those checks run again.

So in the end we have 'duplicated' code, but it's more performant. do I miss something?

Reply:
Checked insert/update. You are right. What we can consider as an option in the future is to refactor stuff a bit and move edge cases handling logic to separate helper methods - one for arrays and the second one for objects.

Reply:
yes, there is room for improvement!

Reply:
> Checked insert/update. You are right. What we can consider as an option in the future is to refactor stuff a bit and move edge cases handling logic to separate helper methods - one for arrays and the second one for objects.

That sounds nice!! And the code would be way easier to understand.